plan.go

```diff
@@ -71,40 +83,108 @@ func (p *plan) Run(ctx context.Context, initData ...interface{}) (Result, error)
 		return nil, ErrInitialDataMissing
 	}
-	return dataMap, p.run(ctx, dataMap)
+	return dataMap, p.run(ctx, workers, dataMap)
 }
 
-func (p *plan) run(ctx context.Context, dataMap map[string]interface{}) error {
-	for i := range p.order {
-		b := p.order[i]
+type work struct {
+	out     chan<- output
+	wg      *sync.WaitGroup
+	builder *builder
+	dataMap map[string]interface{}
+}
+
+type output struct {
+	outputs []reflect.Value
+	builder *builder
+	err     error
+}
+
+func worker(ctx context.Context, wChan <-chan work) {
+	for w := range wChan {
+		processWork(ctx, w)
+	}
+}
+
+func processWork(ctx context.Context, w work) {
+	defer w.wg.Done() // ensure we close wait group
+	o := output{builder: w.builder}
+	fn := reflect.ValueOf(w.builder.Builder)
+	args := make([]reflect.Value, 1)
+	args[0] = reflect.ValueOf(ctx) // first arg is context.Context
+	for _, in := range w.builder.In {
+		data, ok := w.dataMap[in]
+		if !ok {
+			o.err = ErrWTF
+			w.out <- o
+			return
+		}
+		args = append(args, reflect.ValueOf(data))
+	}
+	o.outputs = fn.Call(args)
+	w.out <- o
+}
+
+func doWorkAndGetResult(ctx context.Context, builders []*builder, dataMap map[string]interface{}, wChan chan<- work) error {
+	// create a output channel to read results
+	outChan := make(chan output, len(builders)+1)
+	// create a wait group to wait for all results
+	var wg sync.WaitGroup
+	ctx = AddResultToCtx(ctx, dataMap) // allow builders to access already built data
+	for j := range builders {
+		b := builders[j]
 		if _, ok := dataMap[b.Out]; ok {
 			// do not run the builder if the data already exists
 			continue
 		}
-		v := reflect.ValueOf(b.Builder)
-		input := make([]reflect.Value, 1)
-		ctx = AddResultToCtx(ctx, dataMap) // allow builders to access already built data
-		input[0] = reflect.ValueOf(ctx)
-		for _, in := range b.In {
-			data, ok := dataMap[in]
-			if !ok {
-				return errors.New("What a Terrible Failure!, This is likely a bug in dependency resolution, please report this :|")
-			}
-			input = append(input, reflect.ValueOf(data))
+		// build work
+		w := work{}
+		w.builder = b
+		w.wg = &wg
+		w.dataMap = dataMap
+		w.out = outChan
+		wg.Add(1)  // increment count
+		wChan <- w // send work to be done by workers
+	}
+	wg.Wait() // wait for work to be processed
+	close(outChan)
+	for o := range outChan {
+		if o.err != nil {
+			return o.err
 		}
-		outputs := v.Call(input)
+		outputs := o.outputs
 		// we should only ever have two outputs
 		// 0-> data, 1-> error
 		if !outputs[1].IsNil() {
 			// error occured, return it back and stop processing
 			return outputs[1].Interface().(error)
 		}
+		// add result
 		name := getStructName(outputs[0].Type())
 		dataMap[name] = outputs[0].Interface()
 	}
 	return nil
 }
 
+func (p *plan) run(ctx context.Context, workers uint, dataMap map[string]interface{}) error {
+	if workers == 0 {
+		workers = 1
+	}
+
+	// create a work channel and start workers
+	wChan := make(chan work, 0)
```

Comment:
do we need to handle closing of this channel?

Reply:
thats a bug haha, this is what is causing go-routine leak, let me fix it and add a check for go-routine